### PR TITLE
Call bootstrap with the routes/steps and custom, app settings to build a form and start the service

### PR DESCRIPTION
**Pull request by JoeChapman**
@UKHomeOffice/hof 

- Add initial bootstrap function and parts
- Add tests
- Add dotfiles

--------------------------------
Please feel free criticise this first attempt at a bootstrap function. The README provides some documentation. If unsure look at the tests.

Contributions very welcome! This should be collaborative.

Still TODO:
- Customising session management
.... much more.

### CODE REVIEW

**Comment by JoeChapman** [comment]
What do people (@easternbloc, @joefitter @gxxm @daniel-ac-martin) think about the name?
Is `bootstrap` a name that will be understood by all users, i.e. interaction designers, research people?

File: package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "express-partial-templates": "^0.1.0",
+    "express-session": "^1.13.0",
+    "hogan-express-strict": "^0.5.4",
+    "redis": "^2.6.0-2"
+  },
+  "devDependencies": {
+    "chai": "^3.5.0",
+    "eslint": "^0.23.0",
+    "eslint-config-homeoffice": "^0.1.4",
+    "hof": "^7.0.0",
```
Comment (bendiggle) [inline package.json]: This version of hof created an error when installing the dependancies. 
```
Error: Cannot find module 'hogan.js'
```

File: index.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const app = require('express')();
+// const churchill = require('churchill');
```
Comment (bendiggle) [inline index.js]: Remove comment

File: lib/defaults.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+const hof = require('hof');
+
+/* eslint no-process-env:0 */
+let caller = process.env.NODE_ENV === 'test' ? 'test/integration/fixtures' : module.parent.filename;
```
Comment (bendiggle) [inline lib/defaults.js]: Should this not be a `const` ?

File: test/integration/index.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        strap.server.should.be.an.instanceof(http.Server)
+      )
+    );
+
+    it('should start an Express server', () =>
+      promise.then(strap =>
+        request(strap.server)
+          .get('/one')
+          .expect(200)
+
```
Comment (bendiggle) [inline test/integration/index.js]: remove whitespace

**Comment by JoeChapman** [comment]
when testing this with a service (GRO), errors arise when hitting the first route. 
The app starts, but when the initial route is visited, a TypeError is thrown inside node crypto.
Trial and error suggests that this is not a problem in the implementation, but an issue with the location of the applications' entry point, which is now in /node_modules.

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# EditorConfig is awesome: http://EditorConfig.org
```
Comment (easternbloc) [inline .editorconfig]: not sure this should exist

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+before_install:
+  - rvm install 2.2.2
+
+before_script:
+  - npm run sass
+  - NODE_ENV='ci' npm start&
+
+script:
+  - npm run test:ci
+  - npm run test:acceptance
```
Comment (easternbloc) [inline .travis.yml]: delete this line I think

File: index.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+    if (this.config.getTerms === true) {
+      app.get('/terms-and-conditions', (req, res) => res.render('terms'));
+    }
+
+    bootstrap.use(this.config.errorHandler);
+
+    return bootstrap.start();
+
+  }).call(null, options);
```
Comment (danielepolencic) [inline index.js]: since the context is not used (`null`) you can just unwrap the function and return `bootstrap.start()`.

File: lib/sessions.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  let secure;
+
+  app.use((req, res, next) => {
+    secure = req.protocol === 'https' ? true : false;
+    next();
+  })
+
+  if (secure) {
+    app.set('trust proxy', 1);
+  }
```
Comment (danielepolencic) [inline lib/sessions.js]: is this code executed at all? `secure` is defined only at the first request, whereas its value is used here when the file is loaded.

File: lib/sessions.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  app.use(cookieParser(config.session.secret));
+  app.use((req, res, next) => {
+    const cookie = res.cookie.bind(res);
+    res.cookie = (name, value, options) => {
+      options = options || {};
+      options.secure = secure;
+      options.httpOnly = true;
+      options.path = '/';
+      cookie(name, value, options);
```
Comment (danielepolencic) [inline lib/sessions.js]: all the options can be specified as a second argument in the `cookieParser` method. there's no need to rewrap and rebind `res.cookie`.

```javascript
app.use(cookieParser(config.session.secret, {secure: true, httpOnly: true}))
```

File: lib/settings.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  const viewEngine = config.viewEngine || 'html';
+
+  app.use((req, res, next) => {
+    req.baseUrl = config.siteroot + req.baseUrl;
+    res.locals.assetPath = config.assets;
+    res.locals.gaTagId = config.ga && config.ga.tagId;
+    next();
+  });
+  app.use((req, res, next) => {
+    if (req.baseUrl === 'undefined') {
```
Comment (danielepolencic) [inline lib/settings.js]: why `'undefined'` and not `undefined`?

Comment (JoeChapman) [inline lib/settings.js]: I haven't yet worked out why baseUrl is being set to 'undefined', but unfortunately it is.

File: lib/sessions.js (hunk 8 above)
Comment (JoeChapman) [inline lib/sessions.js]: good point

**Comment by danielepolencic** [comment]
I'd suggest we merge this PR.

It's very big and very hard to contribute any further.

Any objection?

**Comment by joefitter** [comment]
I was going to suggest removing .dockerignore as docker isn't really a 'thing' at this level. But LGTM other than that

**Comment by JoeChapman** [comment]
No objection from me ;)

**Comment by JoeChapman** [comment]
I think you are right about .dockerignore, there being no Dockerfile in the project makes .dockerignore totes redundant

**Comment by joefitter** [comment]
Lets do it

**Comment by JoeChapman** [comment]
@danielepolencic @joefitter I need one of you to merge this please, if you are happy with it thus far

**Comment by joefitter** [comment]
👍

**Pull request merged**